feat: Add support for environments

rlmartin opened the pull request:
Fixes #607

I tested this manually (using a private GitHub App + [test repo](https://github.com/rlmartin/test-repository-settings-app) + smee.io) and it works.

travi commented:
you've done the steps that you need to. i just need to find time to review. this looks like a great contribution, so thank you for the effort you've put in. i'm hopeful that i can take a look by friday afternoon.

rlmartin commented:
👍 thanks so much

travi commented:
i did not get a chance to review this today, but still have it high on my list, so i hope to get to it soon

rlmartin commented:
Hi @travi - do you think you'll get a chance to review this soon? Thanks!

rlmartin commented:
FYI I just added a second commit because I realized that setting `deployment_branch_policy.custom_branch_policies = true` needs to be combined with setting a list of branches using the [deployment branch policies API](https://docs.github.com/en/rest/deployments/branch-policies?apiVersion=2022-11-28). I updated the tests (which are passing) and manually tested again (using smee.io).

rlmartin commented:
FYI after this is merged, the GitHub App settings will need to be updated to request `Read and write` access on `environments` in Repository permissions.

rlmartin commented:
Hi @travi - just checking in on a review. Thanks!

rlmartin commented:
Hi @travi I would really appreciate a review here, before this PR gets too stale and I risk possibly needing to make larger fixes to accommodate any change that has happened in the interim. Thanks!

travi commented:
sorry for the delay getting back to this. we've been working to get some releases out the door for semantic-release, so that work has taken my focus recently.

> the GitHub App settings will need to be updated to request `Read and write` access on `environments` in Repository permissions.

do you have a source for this detail? from https://docs.github.com/en/rest/deployments/environments?apiVersion=2022-11-28#create-or-update-an-environment, it looks like `administration:write` is needed (which settings already asks for), but i don't see a requirement for additional permissions beyond that. Also, I was able to test locally and have it work an app definition that does not have any environments specific permissions. Maybe this has changed?

travi commented on README.md:
```diff
@@ -116,6 +116,24 @@ collaborators:
   # * `maintain` - Recommended for project managers who need to manage the repository without access to sensitive or destructive actions.
   # * `triage` - Recommended for contributors who need to proactively manage issues and pull requests without write access.
 
+# See https://docs.github.com/en/rest/deployments/environments#create-or-update-an-environment for available options
+# Note: deployment_branch_policy differs from the API for ease of use. Either protected_branches (boolean) OR custom_branches (array of strings) is required; this will manage the API requirements under the hood. See https://docs.github.com/en/rest/deployments/branch-policies for documentation of custom_branches. If both are provided in an unexpected manner, protected_branches will be used.
```
https://docs.github.com/en/rest/deployments/environments?apiVersion=2022-11-28#create-or-update-an-environment shows an option to allow all branches by setting the `deployment_branch_policy` to `null`. is that enabled with this approach?

travi replied:
@rlmartin not sure if you saw this question. i think the rest of the PR looks good, but i think we should handle this if it isnt already.

rlmartin replied:
Good question. I'll take a look and get back to you.

rlmartin replied:
Good catch - subtle difference between `undefined` and `null` in the GitHub API. I fixed it now and manually verified that it works (and updated the documentation here).

rlmartin commented:
Oh, hmm, I just kind of assumed it was required, since it is there in the list of permissions when setting up the app (which I did to test this). If it works with _only_ `administration:write`, great - but I didn't see `environments` listed there at all.
![Screenshot 2023-01-27 at 6 16 18 PM](https://user-images.githubusercontent.com/1110255/215223735-f55b818c-793b-4035-8f28-6fb2175c23a6.png)

rlmartin commented:
@travi I removed the `environments:write` permission from my test app and it still worked. So I agree that there is no need to update the App's requested settings - `administration:write` appears to be sufficient. Not sure why there is that extra `environments:write` there 🤷

rlmartin commented:
@travi I believe I have addressed both of your comments and this is ready for re-review.

travi commented on README.md:
```diff
@@ -116,6 +116,25 @@ collaborators:
   # * `triage` - Recommended for contributors who need to proactively manage issues and pull requests without write access.
 
+# See https://docs.github.com/en/rest/deployments/environments#create-or-update-an-environment for available options
+# Note: deployment_branch_policy differs from the API for ease of use. Either protected_branches (boolean) OR custom_branches (array of strings) is required; this will manage the API requirements under the hood. See https://docs.github.com/en/rest/deployments/branch-policies for documentation of custom_branches. If both are provided in an unexpected manner, protected_branches will be used.
```
> (and updated the documentation here)

i think we should make this small adjustment now that null is allowed

```suggestion
# Note: deployment_branch_policy differs from the API for ease of use. Either protected_branches (boolean) OR custom_branches (array of strings) can be provided; this will manage the API requirements under the hood. See https://docs.github.com/en/rest/deployments/branch-policies for documentation of custom_branches. If both are provided in an unexpected manner, protected_branches will be used.
```

rlmartin replied:
Sounds good. Yes setting `deployment_branch_policy: null` will result in `All branches`.

travi reviewed (changes requested):
one more small tweak to the docs seems worthwhile. i think we are ready to merge with that change

travi reviewed (approved):
thanks a lot for this contribution, @rlmartin!

rlmartin commented:
@travi thank you!
